File: src/Appwrite/Network/Validator/Origin.php
<?php

namespace Appwrite\Network\Validator;

use Appwrite\Network\Platform;
use Utopia\Validator;
use Utopia\Validator\Hostname;

class Origin extends Validator
{
    protected ?string $scheme = null;
    protected ?string $host = null;
    protected string $origin = '';

    /**
     * Constructor
     *
     * @param array<string> $allowedHostnames
     * @param array<string> $allowedSchemes
     */
    public function __construct(protected array $allowedHostnames, protected array $allowedSchemes)
    {
    }


    /**
     * Check if Origin is valid.
     * @param mixed $origin The Origin URI.
     * @return bool
     */
    public function isValid($origin): bool
    {
        $this->origin = $origin;
        $this->scheme = null;
        $this->host = null;

        if (!is_string($origin) || empty($origin)) {
            return false;
        }

        $this->scheme = $this->parseScheme($origin);
        $this->host = strtolower(parse_url($origin, PHP_URL_HOST) ?? '');

        $webPlatforms = [
            Platform::SCHEME_HTTP,
            Platform::SCHEME_HTTPS,
            Platform::SCHEME_CHROME_EXTENSION,
            Platform::SCHEME_FIREFOX_EXTENSION,
            Platform::SCHEME_SAFARI_EXTENSION,
            Platform::SCHEME_EDGE_EXTENSION,
        ];
        if (in_array($this->scheme, $webPlatforms, true)) {
            $validator = new Hostname($this->allowedHostnames);
            return $validator->isValid($this->host);
        }

        if (!empty($this->scheme) && in_array($this->scheme, $this->allowedSchemes, true)) {
            return true;
        }

        return false;
    }

    /**
     * Get Description
     * @return string
     */
    public function getDescription(): string
    {
        $platform = $this->scheme ? Platform::getNameByScheme($this->scheme) : '';
        $host = $this->host ? '(' . $this->host . ')' : '';

        if (empty($this->host) && empty($this->scheme)) {
            return 'Invalid Origin.';
        }

        if (empty($platform)) {
            return 'Invalid Scheme. The scheme used (' . $this->scheme . ') in the Origin (' . $this->origin . ') is not supported. If you are using a custom scheme, please change it to `appwrite-callback-<PROJECT_ID>`';
        }

        return 'Invalid Origin. Register your new client ' . $host . ' as a new '
            . $platform . ' platform on your project console dashboard';
    }

    /**
     * Is array
     * @return bool
     */
    public function isArray(): bool
    {
        return false;
    }

    /**
     * Get Type
     * @return string
     */
    public function getType(): string
    {
        return self::TYPE_STRING;
    }

    /**
     * Parses the scheme from a URI string.
     *
     * @param string $uri The URI string to parse.
     * @return string|null The extracted scheme string (e.g., "http", "exp", "mailto")
     */
    public function parseScheme(string $uri): ?string
    {
        $uri = trim($uri);
        if ($uri === '') {
            return null; // No scheme in empty string
        }

        $scheme = parse_url($uri, PHP_URL_SCHEME);
        if ($scheme === false) {
            if (preg_match('/^([a-z][a-z0-9+.-]*):/i', $uri, $matches)) {
                return $matches[1];
            } else {
                return null;
            }
        } else {
            return $scheme;
        }
    }
}

File: 3.x.x/crates-appwrite/network/src/origin.rs
//! CSRF origin allow-list. Rust port of `Appwrite\Network\Validator\Origin`.

use crate::hostname::HostnameList;
use crate::platform::{
    get_name_by_scheme, LOOPBACK_ALIASES, LOOPBACK_HOSTNAME, SCHEME_CHROME_EXTENSION,
    SCHEME_EDGE_EXTENSION, SCHEME_FIREFOX_EXTENSION, SCHEME_HTTP, SCHEME_HTTPS,
    SCHEME_SAFARI_EXTENSION, SCHEME_TAURI,
};
use crate::url::url_host;

const WEB_PLATFORMS: &[&str] = &[
    SCHEME_HTTP,
    SCHEME_HTTPS,
    SCHEME_CHROME_EXTENSION,
    SCHEME_FIREFOX_EXTENSION,
    SCHEME_SAFARI_EXTENSION,
    SCHEME_EDGE_EXTENSION,
    SCHEME_TAURI,
];

/// PHP `Appwrite\Network\Validator\Origin`.
#[derive(Debug, Clone)]
pub struct Origin {
    allowed_hostnames: Vec<String>,
    allowed_schemes: Vec<String>,
    scheme: Option<String>,
    host: Option<String>,
    origin: String,
}

impl Origin {
    #[must_use]
    pub fn new(
        allowed_hostnames: impl IntoIterator<Item = impl Into<String>>,
        allowed_schemes: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            allowed_hostnames: allowed_hostnames.into_iter().map(Into::into).collect(),
            allowed_schemes: allowed_schemes.into_iter().map(Into::into).collect(),
            scheme: None,
            host: None,
            origin: String::new(),
        }
    }

    pub fn set_allowed_hostnames(
        &mut self,
        allowed_hostnames: impl IntoIterator<Item = impl Into<String>>,
    ) -> &mut Self {
        self.allowed_hostnames = allowed_hostnames.into_iter().map(Into::into).collect();
        self
    }

    pub fn set_allowed_schemes(
        &mut self,
        allowed_schemes: impl IntoIterator<Item = impl Into<String>>,
    ) -> &mut Self {
        self.allowed_schemes = allowed_schemes.into_iter().map(Into::into).collect();
        self
    }

    #[must_use]
    pub fn allowed_hostnames(&self) -> &[String] {
        &self.allowed_hostnames
    }

    #[must_use]
    pub fn allowed_schemes(&self) -> &[String] {
        &self.allowed_schemes
    }

    /// PHP `Origin::isValid()`.
    #[must_use]
    pub fn is_valid(&mut self, origin: &str) -> bool {
        if origin.is_empty() {
            return false;
        }
        self.origin = origin.to_string();
        self.scheme = parse_scheme(origin);
        self.host = url_host(origin).map(|h| h.to_ascii_lowercase());

        if let Some(scheme) = &self.scheme {
            if WEB_PLATFORMS.contains(&scheme.as_str()) {
                let host = self.host.clone().unwrap_or_default();
                let validator = HostnameList::new(self.allowed_hostnames.clone());
                if validator.is_valid(&host) {
                    return true;
                }
                return LOOPBACK_ALIASES.contains(&host.as_str())
                    && validator.is_valid(LOOPBACK_HOSTNAME);
            }
        }

        if let Some(scheme) = &self.scheme {
            if !scheme.is_empty() && self.allowed_schemes.iter().any(|s| s == scheme) {
                return true;
            }
        }
        false
    }

    /// PHP `Origin::getDescription()`.
    #[must_use]
    pub fn description(&self) -> String {
        let platform = get_name_by_scheme(self.scheme.as_deref());
        let host = self
            .host
            .as_deref()
            .filter(|h| !h.is_empty())
            .map_or(String::new(), |h| format!("({h})"));

        if self.host.as_deref().unwrap_or("").is_empty() && self.scheme.is_none() {
            return "Invalid Origin.".into();
        }

        if platform.is_empty() {
            let scheme = self.scheme.as_deref().unwrap_or("");
            return format!(
                "Invalid Scheme. The scheme used ({scheme}) in the Origin ({origin}) is not supported. If you are using a custom scheme, please change it to `appwrite-callback-<PROJECT_ID>`",
                origin = self.origin
            );
        }

        format!(
            "Invalid Origin. Register your new client {host} as a new {platform} platform on your project console dashboard"
        )
    }
}

/// PHP `Origin::parseScheme()`.
#[must_use]
pub fn parse_scheme(uri: &str) -> Option<String> {
    let uri = uri.trim();
    if uri.is_empty() {
        return None;
    }
    if let Some((scheme, _)) = uri.split_once(':') {
        if is_scheme_token(scheme) {
            return Some(scheme.to_ascii_lowercase());
        }
    }
    None
}

fn is_scheme_token(scheme: &str) -> bool {
    let mut chars = scheme.chars();
    let Some(first) = chars.next() else {
        return false;
    };
    if !first.is_ascii_alphabetic() {
        return false;
    }
    chars.all(|c| c.is_ascii_alphanumeric() || matches!(c, '+' | '.' | '-'))
}

#[cfg(test)]
mod tests {
    use super::Origin;

    #[test]
    fn values() {
        let mut validator = Origin::new(
            [
                "appwrite.io",
                "appwrite.test",
                "localhost",
                "appwrite.flutter",
            ],
            ["exp", "appwrite-callback-123"],
        );
        assert!(!validator.is_valid(""));
        assert!(!validator.is_valid("/"));
        assert!(validator.is_valid("https://localhost"));
        assert!(validator.is_valid("http://localhost:80"));
        assert!(validator.is_valid("https://appwrite.io"));
        assert!(!validator.is_valid("https://example.com"));
        assert!(validator.is_valid("exp://"));
        assert!(validator.is_valid("appwrite-callback-123://"));
        assert!(!validator.is_valid("appwrite-callback-456://"));
        assert!(!validator.is_valid("appwrite-ios://com.company.appname"));
        assert_eq!(
            validator.description(),
            "Invalid Origin. Register your new client (com.company.appname) as a new iOS platform on your project console dashboard"
        );
        assert!(validator.is_valid("tauri://localhost"));
        assert!(!validator.is_valid("tauri://example.com"));
        assert!(!validator.is_valid("random-scheme://localhost"));
        assert!(validator.description().contains("Invalid Scheme"));
    }

    #[test]
    fn loopback() {
        let mut validator = Origin::new(["appwrite.io", "localhost"], ["exp"]);
        assert!(validator.is_valid("http://localhost"));
        assert!(validator.is_valid("http://localhost:3000"));
        assert!(validator.is_valid("http://127.0.0.1"));
        assert!(validator.is_valid("https://127.0.0.1:5173"));
        assert!(validator.is_valid("http://[::1]"));
        assert!(validator.is_valid("http://[::1]:3000"));
        assert!(!validator.is_valid("http://127.0.0.1.evil.com"));
        assert!(!validator.is_valid("http://localhost.evil.com"));
        assert!(!validator.is_valid("http://128.0.0.1"));
        assert!(!validator.is_valid("http://[2001:db8::1]"));
        assert!(!validator.is_valid("http://xlocalhost"));
        assert!(!validator.is_valid("http://127.0.0.2"));
        assert!(!validator.is_valid("http://[0:0:0:0:0:0:0:1]"));
        assert!(!validator.is_valid("http://localhost."));
        assert!(!validator.is_valid("random-scheme://127.0.0.1"));
    }

    #[test]
    fn loopback_requires_localhost() {
        let mut validator = Origin::new(["appwrite.io"], ["exp"]);
        assert!(!validator.is_valid("http://localhost"));
        assert!(!validator.is_valid("http://127.0.0.1"));
        assert!(!validator.is_valid("http://[::1]"));
    }
}
